README.md
# `singleton-gadts`
[![Haskell Programming Language](https://img.shields.io/badge/language-Haskell-blue.svg)][Haskell.org]
[![BSD3 License](http://img.shields.io/badge/license-BSD3-brightgreen.svg)][tl;dr Legal: BSD3]
[![Build](https://img.shields.io/travis/RyanGlScott/singleton-gadts.svg)](https://travis-ci.org/RyanGlScott/singleton-gadts)

[Haskell.org]:
  http://www.haskell.org
  "The Haskell Programming Language"
[tl;dr Legal: BSD3]:
  https://tldrlegal.com/license/bsd-3-clause-license-%28revised%29
  "BSD 3-Clause License (Revised)"

This package provides a variant of the `SingKind` class from the `singletons` library which supports GADTs and poly-kinded data types, as well as Template Haskell functionality to quickly generate instances of it.

The design of this library is heavily inspired by code written [here](https://github.com/goldfirere/singletons/issues/150#issuecomment-306539091) by Richard Eisenberg. This technique is presented as a separate library (instead of incorporating the technique directly into the `singletons` library) due to some annoying steps one must take to work around [GHC Trac #12088](https://ghc.haskell.org/trac/ghc/ticket/12088), which affects the code here. Refer to the documentation in the `Data.Singletons.GADT` module for more information.


CHANGELOG.md
## 0.1 [????.??.??]
* First version.


tests/Main.hs
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -Wno-orphans #-}
module Main where

import Data.Kind
import Data.Nat
import Data.Singletons.GADT.Prelude ()
import Data.Singletons.GADT.TH

main :: IO ()
main = pure ()

$(singletons1 [d|
  data Foo :: Type -> Type where
    MkFoo :: Foo Bool

  data Quux (a :: Type) where
    MkQuux1 :: Quux Bool
    MkQuux2 :: Quux Ordering

  data Fin :: Nat -> Type where
    FZ :: Fin (S n)
    FS :: Fin n -> Fin (S n)

  data Vec (n :: Nat) (a :: Type) where
    VNil  :: Vec Z a
    VCons :: a -> Vec n a -> Vec (S n) a

  data Prox (a :: k) = P

  data HList :: [Type] -> Type where
    HNil  :: HList '[]
    HCons :: x -> HList xs -> HList (x:xs)
  |])
$(genSingKindInsts2 [''Foo, ''Quux, ''Fin, ''Vec, ''Prox, ''HList])


src/Data/Singletons/GADT.hs
{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE PatternSynonyms #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE ViewPatterns #-}
{-# LANGUAGE UndecidableInstances #-}
{-|
Module:      Data.Singletons.GADT
Copyright:   (C) 2018 Ryan Scott
License:     BSD-style (see the file LICENSE)
Maintainer:  Ryan Scott
Stability:   Experimental
Portability: GHC

Defines a 'SingKind' class (and supporting type families) which can support
GADTs and poly-kinded data types.
-}
module Data.Singletons.GADT (
    -- * Basic definitions
    Promote, PromoteX
  , Demote, DemoteX
  , PromoteDemoteInverse
  , SingKindC, SingKindX, SingKind(..)
    -- $working_around_staging

    -- * Useful combinators
  , withSomeSing
  , singThat
  , demote
  , pattern FromSing

    -- * Defunctionalization symbols
  , DemoteSym0, DemoteSym1
  , PromoteSym0, PromoteSym1
  {-
  , DemoteXSym0, DemoteXSym1
  , PromoteXSym0, PromoteXSym1
  -}
  , PromoteDemoteInverseSym0, PromoteDemoteInverseSym1
  , SingKindCSym0, SingKindCSym1
  , SingKindXSym0, SingKindXSym1

    -- * The rest of "Data.Singletons"
  , module Data.Singletons
  ) where

import Data.Kind
import Data.Singletons hiding
  ( SingKind(..), DemoteSym0, DemoteSym1, FromSing
  , demote, singThat, withSomeSing )
import Data.Singletons.TH (genDefunSymbols)
import Data.Type.Equality
import Unsafe.Coerce

-----
-- Basic defintions
-----

-- | Get a promoted kind from the base type. For example:
--
-- * @'Promote' 'Bool' = 'Bool'@
-- * @'Promote' ('Maybe' a) = 'Maybe' ('Promote' a)@
-- * @'Promote' ('Either' a b) = 'Either' ('Promote' a) ('Promote' b)@
--
-- Rarely, the type and kind do not match. For example:
--
-- * @'Promote' (a -> b) = 'Promote' a '~>' 'Promote' b@
-- * @'Promote' Natural = Nat@
-- * @'Promote' Text = Symbol@
type family Promote (k :: Type) :: Type
-- | A generalized form of 'Promote' which can work over types of any kind.
-- @'PromoteX' (a :: k)@ is @'Promote' a@ when @k@ is @Type@.
-- For some non-@Type@ examples:
--
-- * @'PromoteX' 'True' = 'True'@
-- * @'PromoteX' ('Just' x) = 'Just' ('PromoteX' x)@
-- * @'PromoteX' ((:) a) = 'TyCon1' ((:) ('PromoteX' a))@
type family PromoteX (a :: k) :: Promote k

-- | Get a base type from the promoted kind. For example:
--
-- * @'Demote' 'Bool' = 'Bool'@
-- * @'Demote' ('Maybe' a) = 'Maybe' ('Demote' a)@
-- * @'Demote' ('Either' a b) = 'Either' ('Demote' a) ('Demote' b)@
--
-- Rarely, the type and kind do not match. For example:
--
-- * @'Demote' (a '~>' b) = 'Demote' a -> 'Demote' b@
-- * @'Demote' Nat = Natural@
-- * @'Demote' Symbol = Text@
type family Demote (k :: Type) :: Type
-- | A generalized form of 'Demote' which can work over types of any kind.
-- @'DemoteX' (a :: k)@ is @'Demote' a@ when @k@ is @Type@.
-- For some non-@Type@ examples:
--
-- * @'DemoteX' 'True' = 'True'@
-- * @'DemoteX' ('Just' x) = 'Just' ('DemoteX' x)@
-- * @'DemoteX' ('TyCon1' ((:) a)) = (:) ('DemoteX' a)@
type family DemoteX (a :: k) :: Demote k

-- | A proposition which states that 'PromoteX' and 'DemoteX' are inverses.
-- This is a particularly valuable hint for type inference in typical
-- 'SingKind' instances for GADTs.
type PromoteDemoteInverse (a :: k) = (PromoteX (DemoteX a) ~~ a)

-- | @'SingKindC' a@ is @'SingKind' a@ when @a@ is a 'Type'. If @a@ is not
-- a 'Type', then @'SingKindC' a@ recurs into @a@, gathering up 'SingKind'
-- constraints when necessary. (For instance,
-- @'SingKindC' (x:xs) = ('SingKindC' x, 'SingKindC' xs)@.)
type family SingKindC (a :: k) :: Constraint
-- | A combination of 'SingKindC' and 'PromoteDemoteInverse'. Typically, a
-- 'SingKind' instance will have a 'SingKindX' constraint on each of the
-- type parameters to the data type receiving and instance.
-- (For example,
-- @instance ('SingKindX' a_1, ..., 'SingKindX' a_n) => 'SingKind' (D a_1 ... a_n)@.)
type SingKindX (a :: k) = (SingKindC a, PromoteDemoteInverse a)

-- | The 'SingKind' class is a /kind/ class. It classifies all kinds
-- for which singletons are defined. The class supports converting between a singleton
-- type and the base (unrefined) type which it is built from.
--
-- For a 'SingKind' instance to be well behaved, it should obey the following laws:
--
-- @
-- 'toSing' . 'fromSing' ≡ 'SomeSing'
-- (\\x -> 'withSomeSing' x 'fromSing') ≡ 'id'
-- @
--
-- The final law can also be expressed in terms of the 'FromSing' pattern
-- synonym:
--
-- @
-- (\\('FromSing' sing) -> 'FromSing' sing) ≡ 'id'
-- @
class PromoteDemoteInverse k => SingKind k where
  -- | Convert a singleton to its unrefined version.
  fromSing :: Sing (a :: k) -> Demote k
  -- | Convert an unrefined type to an existentially-quantified singleton type.
  toSing :: Demote k -> SomeSing k

{- $work_around_staging
Due to GHC Trac #12088, a naïve attempt at defining 'Demote'/'Promote' and
'DemoteX'/'PromoteX' instances all in one go will not succeed. For instance,
the following will fail to typecheck:

@
type instance 'Demote'  [a] = ['Demote'  a]
type instance 'Promote' [a] = ['Promote' a]

type instance 'DemoteX' \'[]   = \'[]
type instance 'DemoteX' (x:xs) = 'DemoteX' x : 'DemoteX' xs

type instance 'PromoteX' \'[]   = \'[]
type instance 'PromoteX' (x:xs) = 'PromoteX' x : 'PromoteX' xs
@

The errors will probably look something to the effect of:

@
    • Expected kind ‘Demote [k0]’,
        but ‘DemoteX x : DemoteX xs’ has kind ‘[Demote k0]’
    • In the type ‘DemoteX x : DemoteX xs’
      In the type instance declaration for ‘DemoteX’
@

Because of Trac #12088, the @'Demote' [k0]@ in the kind of the
@'DemoteX' (x:xs)@ instance does not reduce properly. Don't despair, though:
one can work around this issue by explicitly separating the two groups of
declarations using a Template Haskell splice:

@
&#123;-&#35; LANGUAGE TemplateHaskell &#35;-&#125;

type instance 'Demote'  [a] = ['Demote'  a]
type instance 'Promote' [a] = ['Promote' a]

$(return [])

type instance 'DemoteX' \'[]   = \'[]
type instance 'DemoteX' (x:xs) = 'DemoteX' x : 'DemoteX' xs

type instance 'PromoteX' \'[]   = \'[]
type instance 'PromoteX' (x:xs) = 'PromoteX' x : 'PromoteX' xs
@

It's ugly, but it works.
-}

-----
-- Useful combinators
-----

-- | Convert a normal datatype (like 'Bool') to a singleton for that datatype,
-- passing it into a continuation.
withSomeSing :: forall k r
              . SingKind k
             => Demote k                          -- ^ The original datatype
             -> (forall (a :: k). Sing a -> r)    -- ^ Function expecting a singleton
             -> r
withSomeSing x f =
  case toSing x of
    SomeSing x' -> f x'

-- | A convenience function that names a singleton satisfying a certain
-- property.  If the singleton does not satisfy the property, then the function
-- returns 'Nothing'. The property is expressed in terms of the underlying
-- representation of the singleton.
singThat :: forall (a :: k). (SingKind k, SingI a)
         => (Demote k -> Bool) -> Maybe (Sing a)
singThat p = withSing $ \x -> if p (fromSing x) then Just x else Nothing

-- | A convenience function that takes a type as input and demotes it to its
-- value-level counterpart as output. This uses 'SingKind' and 'SingI' behind
-- the scenes, so @'demote' = 'fromSing' 'sing'@.
--
-- This function is intended to be used with @TypeApplications@. For example:
--
-- >>> demote @True
-- True
--
-- >>> demote @(Nothing :: Maybe Ordering)
-- Nothing
demote :: forall a. (SingKind (KindOf a), SingI a) => Demote (KindOf a)
demote = fromSing (sing @a)

-- | An explicitly bidirectional pattern synonym for going between a
-- singleton and the corresponding demoted term.
--
-- As an __expression__: this takes a singleton to its demoted (base)
-- type.
--
-- >>> :t FromSing \@Bool
-- FromSing \@Bool :: Sing a -> Bool
-- >>> FromSing SFalse
-- False
--
-- As a __pattern__: It extracts a singleton from its demoted (base)
-- type.
--
-- @
-- singAnd :: 'Bool' -> 'Bool' -> 'SomeSing' 'Bool'
-- singAnd ('FromSing' singBool1) ('FromSing' singBool2) =
--   'SomeSing' (singBool1 %&& singBool2)
-- @
--
-- instead of writing it with 'withSomeSing':
--
-- @
-- singAnd bool1 bool2 =
--   'withSomeSing' bool1 $ \singBool1 ->
--     'withSomeSing' bool2 $ \singBool2 ->
--       'SomeSing' (singBool1 %&& singBool2)
-- @
pattern FromSing :: SingKind k => forall (a :: k). Sing a -> Demote k
pattern FromSing sng <- ((\demotedVal -> withSomeSing demotedVal SomeSing) -> SomeSing sng)
  where FromSing sng = fromSing sng

-----
-- Particularly important instances
-----

type instance Demote  Type = Type
type instance Promote Type = Type
type instance DemoteX   (a :: Type) = Demote a
type instance PromoteX  (a :: Type) = Promote a
type instance SingKindC (a :: Type) = SingKind a

type instance Demote  (a ~> b) = DemoteX  a -> DemoteX  b
type instance Promote (a -> b) = PromoteX a ~> PromoteX b
instance (SingKindX k1, SingKindX k2) => SingKind (k1 ~> k2) where
  fromSing sFun x = withSomeSing x $ fromSing . applySing sFun
  toSing f = SomeSing slam
    where
      slam :: forall (f :: k1 ~> k2). Sing f
      slam = singFun1 @f lam
        where
          lam :: forall (t :: k1). Sing t -> Sing (f @@ t)
          lam x = withSomeSing (f (fromSing x)) (\(r :: Sing res) -> unsafeCoerce r)

-----
-- Defunctionalization symbols
-----

$(genDefunSymbols [ ''Demote , ''Promote
                  -- Can't do these yet due to https://ghc.haskell.org/trac/ghc/ticket/12564
                  -- , ''DemoteX, ''PromoteX
                  , ''PromoteDemoteInverse, ''SingKindC, ''SingKindX
                  ])


src/Data/Singletons/GADT/Prelude.hs
{-# LANGUAGE EmptyCase #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -Wno-orphans #-}
{-|
Module:      Data.Singletons.GADT.Prelude
Copyright:   (C) 2018 Ryan Scott
License:     BSD-style (see the file LICENSE)
Maintainer:  Ryan Scott
Stability:   Experimental
Portability: GHC

Defines 'SingKind' instances for many data types, including some types which
cannot be given instances of the @SingKind@ variant offered by the @singletons@
library.
-}
module Data.Singletons.GADT.Prelude (
    Sing(..)
  , type (%:~:), type (%:~~:)

    -- * Defunctionalization symbols
  , ReflSym0, HReflSym0

    -- * The rest of "Data.Singletons.GADT"
  , module Data.Singletons.GADT

    -- * The rest of "Data.Singletons.Prelude"
  , module Data.Singletons.Prelude
  ) where

import           Data.Nat
import           Data.Proxy (Proxy(..))
import           Data.Singletons.GADT
import           Data.Singletons.GADT.Prelude.Internal
import           Data.Singletons.GADT.TH
import           Data.Singletons.Prelude hiding ( SingKind(..), DemoteSym0, DemoteSym1
                                                , FromSing, demote, singThat, withSomeSing )
import           Data.Singletons.Prelude.Const (Sing(..))
import           Data.Singletons.Prelude.Identity (Sing(..))
import           Data.Singletons.Prelude.List.NonEmpty (Sing(..))
import           Data.Singletons.Prelude.Monoid hiding (Sing(SFirst, SLast, sGetFirst, sGetLast))
import           Data.Singletons.Prelude.Semigroup hiding (Sing(SFirst, SLast, sGetFirst, sGetLast))
import           Data.Singletons.TypeLits (Sing(..))
import qualified Data.Text as T
import           Data.Text (Text)

import           GHC.TypeLits (SomeSymbol(..), Symbol, someSymbolVal, symbolVal)
import qualified GHC.TypeNats as TN
import           GHC.TypeNats (SomeNat(..))

import           Numeric.Natural (Natural)

$(genSingKindInsts1 existingSingInstNames)
$(genSingKindInsts2 existingSingInstNames)

type instance Demote TN.Nat = Natural
type instance Promote Natural = TN.Nat
type instance SingKindC (n :: TN.Nat) = ()
instance SingKind TN.Nat where
  fromSing (SNat :: Sing n) = TN.natVal (Proxy :: Proxy n)
  toSing n = case TN.someNatVal n of
               SomeNat (_ :: Proxy n) -> SomeSing (SNat :: Sing n)

type instance Demote Symbol = Text
type instance Promote Text = Symbol
type instance SingKindC (s :: Symbol) = ()
instance SingKind Symbol where
  fromSing (SSym :: Sing n) = T.pack (symbolVal (Proxy :: Proxy n))
  toSing s = case someSymbolVal (T.unpack s) of
               SomeSymbol (_ :: Proxy n) -> SomeSing (SSym :: Sing n)

-- Time to show off
$(genSingletons1    newSingInstNames)
$(genSingKindInsts2 newSingInstNames)


src/Data/Singletons/GADT/TH.hs
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-|
Module:      Data.Singletons.GADT.Prelude
Copyright:   (C) 2018 Ryan Scott
License:     BSD-style (see the file LICENSE)
Maintainer:  Ryan Scott
Stability:   Experimental
Portability: GHC

Defines Template Haskell functions for quickly generating instances of
'SingKind' (and supporting type families).
-}
module Data.Singletons.GADT.TH (
    -- $working_around_staging
    -- * Reimplementations of functionality from "Data.Singletons.TH"
    genSingletons
  , genSingletons1
  , singletons
  , singletons1
  , singletonsOnly
  , singletonsOnly1

    -- * Generating instances of classes/families in "Data.Singletons.GADT"
  , genSingKindInsts
  , genSingKindInsts1
  , genSingKindInsts2

    -- * "Data.Singletons.GADT" reexports
  , module Data.Singletons.GADT

    -- * The rest of "Data.Singletons.TH"
  , module Data.Singletons.TH
  ) where

import           Control.Monad

import           Data.Foldable
import           Data.Maybe
import qualified Data.Singletons as Old (SingKind)
import           Data.Singletons.GADT
import           Data.Singletons.Prelude (Sing(..))
import qualified Data.Singletons.TH as TH
import           Data.Singletons.TH hiding
                   ( genSingletons, singletons, singletonsOnly
                   , SingKind(..), DemoteSym0, DemoteSym1, FromSing
                   , demote, singThat, withSomeSing )
import           Language.Haskell.TH.Syntax
import           Language.Haskell.TH.Desugar

{- $working_around_staging
As "Data.Singletons.GADT" explains, one often has to work around issues
caused by GHC Trac #12088. To make this task somewhat easier, the functions
in "Data.Singletons.GADT.TH" all have variants which only generate certain
subsets of instances so that one can predictably sequence Template Haskell
splices in the right order to avoid #12088.

This is all better explained by example. If you wanted to generate the
following instances using Template Haskell.

@
&#123;-&#35; LANGUAGE TemplateHaskell &#35;-&#125;

type instance 'Demote'  [a] = ['Demote'  a]
type instance 'Promote' [a] = ['Promote' a]

$(return [])

type instance 'DemoteX' \'[]   = \'[]
type instance 'DemoteX' (x:xs) = 'DemoteX' x : 'DemoteX' xs

type instance 'PromoteX' \'[]   = \'[]
type instance 'PromoteX' (x:xs) = 'PromoteX' x : 'PromoteX' xs
@

Then one should /not/ do the following:

@
&#123;-&#35; LANGUAGE TemplateHaskell &#35;-&#125;
$('genSingletons' [''[]])
@

As this will generate all of the above instances at once, without any explicit
separation. Instead, one should do the following:

@
&#123;-&#35; LANGUAGE TemplateHaskell &#35;-&#125;
$('genSingletons1' [''[]])
$('genSingletons2' [''[]])
@

'genSingletons1' will first generate the 'Demote'/'Promote' instances, within
a single declaration group, and then the next declaration group generated by
'genSingletons2' will contain the 'DemoteX'/'PromoteX' instances.

'genSingletons' is still provided in the event that Trac #12088 gets fixed one
day, but chances are, it won't be very useful to you.
-}

-----
-- Reimplementations of Data.Singletons.TH functionality
-----

-- | Generate singleton definitions from a type that is already defined.
-- For example, the singletons package itself uses
--
-- > $(genSingletons [''Bool, ''Maybe, ''Either, ''[]])
--
-- to generate singletons for Prelude types.
genSingletons :: DsMonad q => [Name] -> q [Dec]
genSingletons names = do
  (decs1, decs2) <- genSingletons' names
  pure $ decs1 ++ decs2

-- | Like 'genSingletons', but only generates 'Demote', 'Promote', and
-- 'SingKind' instances.
genSingletons1 :: DsMonad q => [Name] -> q [Dec]
genSingletons1 = fmap fst . genSingletons'

genSingletons' :: DsMonad q => [Name] -> q ([Dec], [Dec])
genSingletons' names = do
  oldDecs <- filterOutOldSingKind <$> TH.genSingletons names
  (newDecs1, newDecs2) <- genSingKindInsts' names
  pure (oldDecs ++ newDecs1, newDecs2)

-- | Make promoted and singleton versions of all declarations given, retaining
-- the original declarations.
-- See <https://github.com/goldfirere/singletons/blob/master/README.md> for
-- further explanation.
singletons :: DsMonad q => q [Dec] -> q [Dec]
singletons qdecs = do
  (decs1, decs2) <- singletons' TH.singletons qdecs
  pure $ decs1 ++ decs2

-- | Like 'singletons', but only generates 'Demote', 'Promote', and
-- 'SingKind' instances.
singletons1 :: DsMonad q => q [Dec] -> q [Dec]
singletons1 = fmap fst . singletons' TH.singletons

-- | Make promoted and singleton versions of all declarations given, discarding
-- the original declarations. Note that a singleton based on a datatype needs
-- the original datatype, so this will fail if it sees any datatype declarations.
-- Classes, instances, and functions are all fine.
singletonsOnly :: DsMonad q => q [Dec] -> q [Dec]
singletonsOnly qdecs = do
  (decs1, decs2) <- singletons' TH.singletonsOnly qdecs
  pure $ decs1 ++ decs2

-- | Like 'singletonsOnly', but only generates 'Demote', 'Promote', and
-- 'SingKind' instances.
singletonsOnly1 :: DsMonad q => q [Dec] -> q [Dec]
singletonsOnly1 = fmap fst . singletons' TH.singletonsOnly

singletons' :: DsMonad q => (q [Dec] -> q [Dec]) -> q [Dec] -> q ([Dec], [Dec])
singletons' genSings qdecs = do
  decs <- qdecs
  oldDecs <- filterOutOldSingKind <$> genSings (pure decs)
  (newDecs1, newDecs2) <- singDecs decs
  pure (oldDecs ++ decsToTH newDecs1, decsToTH newDecs2)

-----
-- Macros for generating instances of classes/families from Data.Singletons.GADT
-----

-- | Generate 'Demote', 'Promote', 'SingKind', 'DemoteX', 'PromoteX', and
-- 'SingKindC' instances for each provided data type. Ignores non–data type
-- declarations.
genSingKindInsts :: DsMonad q => [Name] -> q [Dec]
genSingKindInsts names = do
  (decs1, decs2) <- genSingKindInsts' names
  pure $ decs1 ++ decs2

-- | Like 'genSingKindInsts', but only generates 'Demote', 'Promote', and
-- 'SingKind' instances. Ignores non–data type declarations.
genSingKindInsts1 :: DsMonad q => [Name] -> q [Dec]
genSingKindInsts1 = fmap fst . genSingKindInsts'

-- | Like 'genSingKindInsts', but only generates 'DemoteX', 'PromoteX', and
-- 'SingKindC' instances. Ignores non–data type declarations.
genSingKindInsts2 :: DsMonad q => [Name] -> q [Dec]
genSingKindInsts2 = fmap snd . genSingKindInsts'

genSingKindInsts' :: DsMonad q => [Name] -> q ([Dec], [Dec])
genSingKindInsts' names = do
  (decs1, decs2) <- concatMapM (singInfo <=< dsInfo <=< reifyWithLocals) names
  pure (decsToTH decs1, decsToTH decs2)

-----
-- Internals
-----

singDecs :: DsMonad q => [Dec] -> q ([DDec], [DDec])
singDecs decs = do
  ddecs <- withLocalDeclarations decs $ dsDecs decs
  (ddecss1, ddecss2) <- mapAndUnzipM singDec ddecs
  pure (concat ddecss1, concat ddecss2)

singInfo :: DsMonad q => DInfo -> q ([DDec], [DDec])
singInfo (DTyConI dec _) = singDec dec
singInfo _               = pure ([], [])

singDec :: DsMonad q => DDec -> q ([DDec], [DDec])
singDec (DDataD _nd _cxt name tvbs mk cons _derivings)
  = do all_tvbs <- buildDataDTvbs tvbs mk
       singDataD name all_tvbs cons
singDec _ = pure ([], [])

-- Generates all SingKind-related declarations for a given data type,
-- separating the declarations into two groups (as described above) to
-- avoid Template Haskell staging restrictions.
--
-- This is largely cargo-culted from singDataD in the singletons library.
singDataD :: forall q. DsMonad q
          => Name -> [DTyVarBndr] -> [DCon]
          -> q ([DDec], [DDec])
singDataD name tvbs ctors = do
  let tvbNames = map extractTvbName tvbs
      tvbTys   = map DVarT tvbNames
      dataType = foldType (DConT name) tvbTys

  -- Demote, Promote instances
  let mkPromoteDemoteInstance :: Name -> Name -> DDec
      mkPromoteDemoteInstance cls clsX =
        DTySynInstD cls $ DTySynEqn [dataType]
                        $ foldType (DConT name)
                        $ map (DConT clsX `DAppT`) tvbTys

      demoteInstance  = mkPromoteDemoteInstance ''Demote  ''DemoteX
      promoteInstance = mkPromoteDemoteInstance ''Promote ''PromoteX

  -- SingKind instance
  fromSingClauses     <- mapM mkFromSingClause ctors
  emptyFromSingClause <- mkEmptyFromSingClause
  toSingClauses       <- mapM mkToSingClause ctors
  emptyToSingClause   <- mkEmptyToSingClause

  let singKindInstance =
        DInstanceD Nothing
                   (map (DAppPr (DConPr ''SingKindX)) tvbTys)
                   (DAppT (DConT ''SingKind) dataType)
                   [ DLetDec $ DFunD 'fromSing
                             $ fromSingClauses `orIfEmpty` [emptyFromSingClause]
                   , DLetDec $ DFunD 'toSing
                             $ toSingClauses   `orIfEmpty` [emptyToSingClause]
                   ]

  -- DemoteX, PromoteX, SingKindC instances
  -- TODO RGS: Instances for TyConN
  let mkPromoteXDemoteXInstances :: Name -> q [DDec]
      mkPromoteXDemoteXInstances clsX = mkConInstances clsX foldType

      mkSingKindCInstance :: q [DDec]
      mkSingKindCInstance = mkConInstances ''SingKindC (const mkTupleDType)

      mkConInstances :: Name
                     -> (DType -> [DType] -> DType)
                     -> q [DDec]
      mkConInstances cls foldArgs = traverse mkInst ctors
        where
          mkInst :: DCon -> q DDec
          mkInst c = do
            let (cname, numArgs) = extractNameArgs c
            varNames <- replicateM numArgs (qNewName "x")
            let varTys = map DVarT varNames
            pure $ DTySynInstD cls $ DTySynEqn [foldType (DConT cname) varTys]
                                   $ foldArgs (DConT cname)
                                   $ map (DConT cls `DAppT`) varTys

  demoteXInstances   <- mkPromoteXDemoteXInstances ''DemoteX
  promoteXInstances  <- mkPromoteXDemoteXInstances ''PromoteX
  singKindCInstances <- mkSingKindCInstance

  pure ( [demoteInstance, promoteInstance, singKindInstance]
       , demoteXInstances ++ promoteXInstances ++ singKindCInstances
       )
  where
    mkFromSingClause :: DCon -> q DClause
    mkFromSingClause c = do
      let (cname, numArgs) = extractNameArgs c
      varNames <- replicateM numArgs (qNewName "b")
      return $ DClause [DConPa (singDataConName cname) (map DVarPa varNames)]
                       (foldExp
                          (DConE cname)
                          (map (DAppE (DVarE 'fromSing) . DVarE) varNames))

    mkToSingClause :: DCon -> q DClause
    mkToSingClause (DCon _tvbs _cxt cname fields _rty) = do
      let types = tysOfConFields fields
      varNames  <- mapM (const $ qNewName "b") types
      svarNames <- mapM (const $ qNewName "c") types
      -- promoted  <- mapM promoteType types
      let promoted       = types
          varPats        = zipWith mkToSingVarPat varNames promoted
          recursiveCalls = zipWith mkRecursiveCall varNames promoted
      return $
        DClause [DConPa cname varPats]
                (multiCase recursiveCalls
                           (map (DConPa 'SomeSing . pure . DVarPa)
                                svarNames)
                           (DAppE (DConE 'SomeSing)
                                     (foldExp (DConE (singDataConName cname))
                                              (map DVarE svarNames))))

    mkToSingVarPat :: Name -> DKind -> DPat
    mkToSingVarPat varName _ki =
      -- DSigPa (DVarPa varName) (DAppT (DConT ''Demote) ki)
      DVarPa varName

    mkRecursiveCall :: Name -> DKind -> DExp
    mkRecursiveCall var_name _ki =
      {-
      DSigE (DAppE (DVarE 'toSing) (DVarE var_name))
            (DAppT (DConT ''SomeSing) ki)
      -}
      DAppE (DVarE 'toSing) (DVarE var_name)

    mkEmptyFromSingClause :: q DClause
    mkEmptyFromSingClause = do
      x <- qNewName "x"
      pure $ DClause [DVarPa x]
           $ DCaseE (DVarE x) []

    mkEmptyToSingClause :: q DClause
    mkEmptyToSingClause = do
      x <- qNewName "x"
      pure $ DClause [DVarPa x]
           $ DConE 'SomeSing `DAppE` DCaseE (DVarE x) []

-- Discard any type class instances for SingKind (from the singletons library).
filterOutOldSingKind :: [Dec] -> [Dec]
filterOutOldSingKind = filter (not . isSingKind)
  where
    -- If it's stupid but it works, then it ain't stupid.
    isSingKind :: Dec -> Bool
    isSingKind (InstanceD _ _ instTy _)
      | AppT (ConT n) _ <- instTy
      = n == ''Old.SingKind
    isSingKind _ = False

-- | Make a constraint tuple 'DType' from a list of 'DType's. Avoids using a 1-tuple.
mkTupleDType :: [DType] -> DType
mkTupleDType [ty] = ty
mkTupleDType tys  = foldl DAppT (DConT $ tupleTypeName (length tys)) tys

-- Taken from singletons' source code

singDataConName :: Name -> Name
singDataConName nm
  | nm == '[]                                      = 'SNil
  | nm == '(:)                                     = 'SCons
  | Just degree <- tupleNameDegree_maybe nm        = mkTupleDataName degree
  | Just degree <- unboxedTupleNameDegree_maybe nm = mkTupleDataName degree
  | otherwise                                      = prefixConName "S" "%" nm

mkTupleDataName :: Int -> Name
mkTupleDataName n = mkName $ "STuple" ++ (show n)

-- Put an uppercase prefix on a constructor name. Takes two prefixes:
-- one for identifiers and one for symbols.
--
-- This is different from 'prefixName' in that infix constructor names always
-- start with a colon, so we must insert the prefix after the colon in order
-- for the new name to be syntactically valid.
prefixConName :: String -> String -> Name -> Name
prefixConName pre tyPre n = case (nameBase n) of
    (':' : rest) -> mkName (':' : tyPre ++ rest)
    alpha -> mkName (pre ++ alpha)

-- build a pattern match over several expressions, each with only one pattern
multiCase :: [DExp] -> [DPat] -> DExp -> DExp
multiCase [] [] body = body
multiCase scruts pats body =
  DCaseE (mkTupleDExp scruts) [DMatch (mkTupleDPat pats) body]

tysOfConFields :: DConFields -> [DType]
tysOfConFields (DNormalC _ stys) = map snd stys
tysOfConFields (DRecC vstys) = map (\(_,_,ty) -> ty) vstys

-- extract the name and number of arguments to a constructor
extractNameArgs :: DCon -> (Name, Int)
extractNameArgs = fmap length . extractNameTypes

-- extract the name and types of constructor arguments
extractNameTypes :: DCon -> (Name, [DType])
extractNameTypes (DCon _ _ n fields _) = (n, tysOfConFields fields)

-- Construct a data type's variable binders, possibly using fresh variables
-- from the data type's kind signature.
buildDataDTvbs :: DsMonad q => [DTyVarBndr] -> Maybe DKind -> q [DTyVarBndr]
buildDataDTvbs tvbs mk = do
  extra_tvbs <- mkExtraDKindBinders $ fromMaybe (DConT typeKindName) mk
  pure $ tvbs ++ extra_tvbs

-- extract the name from a TyVarBndr.
extractTvbName :: DTyVarBndr -> Name
extractTvbName (DPlainTV n) = n
extractTvbName (DKindedTV n _) = n

-- apply an expression to a list of expressions
foldExp :: DExp -> [DExp] -> DExp
foldExp = foldl DAppE

-- apply a type to a list of types
foldType :: DType -> [DType] -> DType
foldType = foldl DAppT

-- choose the first non-empty list
orIfEmpty :: [a] -> [a] -> [a]
orIfEmpty [] x = x
orIfEmpty x _ = x

-- lift concatMap into a monad
-- could this be more efficient?
concatMapM :: (Monad monad, Monoid monoid, Traversable t)
           => (a -> monad monoid) -> t a -> monad monoid
concatMapM fn list = do
  bss <- mapM fn list
  return $ fold bss
